utils/doclint/generateDotnetApi.js

```diff
@@ -0,0 +1,695 @@
+*/
+function translateType(type, parent, generateNameCallback = null) {
+
+  if (!generateNameCallback) {
+    generateNameCallback = (t) => {
+      return t.name;
+    }
+  }
+
+  // a few special cases we can fix automatically
+  if (type.expression === '[null]|[Error]')
+    return 'void';
+  else if (type.expression === '[boolean]|"mixed"')
+    return 'MixedState';
+
+  let isNullableEnum = false;
+  if (type.union) {
+    if (type.union[0].name === 'null') {
+      // for dotnet, this is a nullable type
+      // if the other side is a primitive type
+      if (type.union.length > 2) {
+        if (type.union.filter(x => x.name.startsWith('"')).length == type.union.length - 1)
+          isNullableEnum = true;
+        else
+          throw `Union (${parent.name}) with null is too long.`;
+      } else {
+        const innerTypeName = translateType(type.union[1], parent, generateNameCallback);
+        // if type is primitive, or an enum, then it's nullable
+        if (innerTypeName === 'bool'
+          || innerTypeName === 'int') {
+          return `${innerTypeName}?`;
+        }
+
+        // if it's not a value type, it'll be nullable by default, so we can ignore it
+        return `${innerTypeName}`;
+      }
+    }
+
+    if (type.union.filter(u => u.name.startsWith(`"`)).length == type.union.length
+      || isNullableEnum) {
+      // this is an enum
+      let enumName = generateNameCallback(type);
+      if (!enumName)
+        throw `This was supposed to be an enum, but it failed generating a name, ${type.name} ${parent ? parent.name : ""}.`;
+
+      // make sure we map the enum, or invalidate the name, in case it doesn't match well
+      const potentialEnum = enumTypes.get(enumName);
+      let enumValues = type.union.filter(x => x.name !== 'null').map(x => x.name);
+      if (potentialEnum) {
+        // compare values
+        if (potentialEnum.join(',') !== enumValues.join(',')) {
+          // for now, we'll merge the two enums, if they have the same name, and we'll go from there
+          potentialEnum.concat(enumValues.filter(x => !potentialEnum.includes(x))); // merge & de-dupe
+          // TODO: think about doing global type annotation, where we can add comments, such as this?
+          enumTypes.set(enumName, potentialEnum);
+        }
+      } else {
+        enumTypes.set(enumName, enumValues);
+      }
+      if (isNullableEnum)
+        return `${enumName}?`;
+      return enumName;
+    }
+
+    if (type.expression === '[string]|[Buffer]')
+      return `byte[]`; // TODO: make sure we implement extension methods for this!
+    else if (type.expression === '[string]|[float]'
+      || type.expression === '[string]|[float]|[boolean]'
+      || (type.union && type.union[0].name === 'string' && type.union[1].name === 'float' && type.union[2].name === 'boolean'))
+      return 'string'; // because users can send either 100 or 100px for most of these arguments
```

Comment:
if this is for Page.pdf.margin all of these have type string for c# and java already

Reply:
Yeah, this was changed as I was working on the generator :-). The question now is, do I leave that in, or assume this will be adhered to as we add stuff... I could just throw an exception and make sure we fix it.